Make at::numeric_limits non CUDA-specific

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50903 Add correction parameter to std/var
* #51350 De-prioritise Dimname and DimnameList in python overload resolution
* **#50902 Make at::numeric_limits non CUDA-specific**

This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

peterbell10 committed:
Make at::numeric_limits non CUDA-specific

This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

mruberry commented on aten/src/ATen/NumericLimits.h:
```diff
@@ -0,0 +1,121 @@
+#include <float.h>
+
+// NumericLimits.h is a holder for numeric limits definitions of commonly used
+// types. This header is very specific to ROCm HIP and may be removed in the
```
cc @ngimel

If this file is really intended only to support ROCm and std::numeric_limits is preferred, should we really make this change?

Follow-up question: has ROCm fixed its std::numeric_limits handling since this was written? Is it still necessary?

ngimel replied:
cc @jeffdaily for ROCm question. ROCm has improved a lot since this was written, it's very possible that it's no longer needed.

peterbell10 replied:
std and var are implemented in `SharedReduceOps.h` which gets included in both CPU and CUDA code. So, this is useful, assuming the file is still necessary for ROCm compilation.

jeffdaily replied:
I'll need a bit of time to confirm the ROCm question.

mruberry replied:
Hey @jeffdaily - do you have an idea for how long this will take to review?

jeffdaily replied:
I would like confirmation if the support of std::numeric_limits is limited to ROCm and not CUDA?  I can't tell whether CUDA supports all features of numeric_limits?
- https://forums.developer.nvidia.com/t/how-to-assign-infinity-to-variables-in-cuda-code/14348
- https://stackoverflow.com/questions/24498140/is-there-a-cuda-equivalent-to-stdnumeric-limits

peterbell10 replied:
Everything in `std::numeric_limits` is constexpr, so at least in theory should work on nvcc in device code when compiled with `--expt-relaxed-constexpr` (which PyTorch is).

mruberry replied:
cc @ptrblck and @zasdfgbnm 

Does CUDA support std::numeric_limits? @colesbury's comment in this file says it does.

I don't want to take much more time on this issue, however. @peterbell10 can we just run the CI with this file removed and std::numeric_limits used and see what happens?

peterbell10 replied:
Seems to be compiling fine when I use `std::numeric_limits` inside of `at::numeric_limits`. However, it occurs to me that the warning is meaningless since the functions in `at::numeric_limits` have different names and meanings from the standard `std::numeric_limits` functions.

Either way, that should mean I don't need the changes in this PR any more.

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]

peterbell10 committed:
Update on "Make at::numeric_limits non CUDA-specific"


This allows numeric_limits to be used in code shared between CPU and CUDA
implementations.

[ghstack-poisoned]
